src/config/mod.rs

```diff
@@ -43,13 +43,21 @@ pub use log_schema::{init_log_schema, log_schema, LogSchema};
 pub use unit_test::build_unit_tests_main as build_unit_tests;
 pub use validation::warnings;
 
+#[derive(Debug, Deserialize, Serialize)]
+pub struct SourceConfigEntry {
+    #[serde(flatten)]
+    pub config: Box<dyn SourceConfig>,
+    #[serde(skip)]
+    pub identifier: Option<Box<str>>,
```

Comment:
It'd be nice to make this non-optional, skip the serde stuff, and make assignment part of the `ConfigBuilder -> Config` process (i.e. the builder field stays `Box<dyn SourceConfig>`), but I'm not sure how much extra work that would be.